Fix for last commit to ActionWait so that both withTimeout and

andrew-sumner opened the pull request:
withMaxAttempts work correctly - and add some tests to prove it.

nigelcharman commented on cubano-core/src/main/java/org/concordion/cubano/driver/action/DurationParser.java:
```diff
@@ -0,0 +1,115 @@
+    /**
+     * A string representation of this duration using ISO-8601 seconds based representation, such as PT8H6M12.345S.
+     * 
+     * The format of the returned string will be PTnHnMnS, where n is the relevant hours, minutes or seconds part of the duration. Any fractional seconds are placed after a decimal
```
Isn't this going the other way around, and returning a description in the English language?

andrew-sumner replied:
Fixed

nigelcharman commented on cubano-core/src/test/java/org/concordion/cubano/driver/action/ActionWaitTests.java:
```diff
@@ -0,0 +1,67 @@
+                return (String) null;
+            });
+
+            throw new IllegalStateException("Should have timed out");
```
Will this ever time out? Consider using https://github.com/junit-team/junit4/wiki/timeout-for-tests.

andrew-sumner replied:
This is a catch incase someone changes something and the wait.until() doesn't time out.  This code should never be reached if the test and real code are working correctly.

andrew-sumner replied:
Code updated and more tests added

nigelcharman replied:
So probably more that the timeout has finished without throwing a TimeoutException. 
I'd still suggest changing the test annotation to something like `@Test(timeout=3000)` in case it doesn't time out

andrew-sumner replied:
The code is now "fail("Test should never reach this point");", unfortunately this view doesn't reflect the code update.

nigelcharman commented on cubano-core/src/test/java/org/concordion/cubano/driver/action/ActionWaitTests.java:
```diff
@@ -0,0 +1,67 @@
+                return (String) null;
+            });
+
+            throw new Exception("Should have timed out");
```
ditto

andrew-sumner replied:
As above

nigelcharman commented on cubano-core/src/test/java/org/concordion/cubano/driver/action/DurationParserTests.java:
```diff
@@ -0,0 +1,44 @@
+        Duration d = Duration.ofDays(1);
+
+        assertThat(DurationParser.toLongString(d), is("24 Hours"));
+    }
```
Would be worth adding more tests to improve coverage of DurationParser.

andrew-sumner replied:
Done

andrew-sumner commented on cubano-core/src/main/java/org/concordion/cubano/driver/action/ActionWait.java:
```diff
@@ -243,15 +243,19 @@ public int getAttempts() {
 
         Instant start = clock.instant();
-        Instant end = null;
+        Instant end = clock.instant();
```
Fixed

nigelcharman replied:
I actually meant `Instant end = start;` to make sure they are identical. As it stands there'll be some difference between the 2 `Instant`s. Since `Instant` is immutable, you'll get a new `Instant` if you change it, so this will be safe.

andrew-sumner replied:
Thats what I did, this thread is showing the older version of the code without my changes - which I personally find rather confusing!  Took me ages to work out my commit had made it.

andrew-sumner commented on cubano-core/src/main/java/org/concordion/cubano/driver/action/ActionWait.java:
```diff
@@ -356,15 +353,17 @@ private int getNextPollingInterval(Instant end) {
+            long endTime = end.toEpochMilli();
+            long currentTime = clock.instant().toEpochMilli();
+            long waitTime = currentTime + TimeUnit.MILLISECONDS.convert(interval, pollingTimeUnit);
+            long stretchTime = waitTime + TimeUnit.MILLISECONDS.convert(interval / 2, pollingTimeUnit);
```
I've updated the comments somewhat, but yes it probably should have some tests around it.  Basically if the wait time for the current polling interval would take it close to the timeout limit then only do the one sleep and final poll rather than sleep, poll, short sleep and final poll.

andrew-sumner replied:
I've added the test "withTimeoutStretchInterval" to illustrate this, hopefully it's not too fragile as I've found differences in running time sensitive tests locally and on travis, even on travis and different java versions.

andrew-sumner commented:
@nigelcharman I think I've addressed all your comments, and I've added some more tests in to improve the coverage.  These tests required some small improvements to the ActionWait class when dealing with millisecond level waits - which is not what this class was designed for but good to fix.

nigelcharman commented:
Will need a merge of master please @andrew-sumner
